middleware/kubernetes/apiproxy.go

```diff
@@ -0,0 +1,194 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type proxyUpstream struct {
+	network   string
+	address   string
+	stop      chan struct{}
+	wait      sync.WaitGroup
+	unhealthy int32
+}
+
+type proxyHandler struct {
+	upstreams []proxyUpstream
+	pickIndex int32
+}
+
+type apiProxy struct {
+	http.Server
+	listener net.Listener
+	handler  proxyHandler
+}
+
+func (p proxyUpstream) Network() string {
+	return p.network
+}
+
+func (p proxyUpstream) String() string {
+	return p.address
+}
+
+func (p *proxyUpstream) Healthcheck(once bool) {
```

Comment:
Did you look at the health checks in middleware/proxy? Looks awfully similar... Either using that directly or getting a new pkg/healthcheck subpackage might sense here.